File: backend/match.ts
#!/usr/bin/env tsx
/**
 * Matching step: extraction_run_id → proposed_changes rows.
 *
 * Resolves po_ref and sku_or_code from LLM output to canonical DB records,
 * then inserts proposed_changes for each resolved line update.
 *
 * Entry point:
 *   CLI: tsx backend/match.ts <extraction_run_uuid>
 */

import pg from 'pg'

const { Pool } = pg
const pool = new Pool({ connectionString: process.env.DATABASE_URL })

// ── types (mirrors the extraction schema) ────────────────────────────────────

interface LineUpdate {
  sku_or_code: string
  field: 'delivery_date' | 'quantity'
  new_value: string
  evidence: string
  confidence: number
}

interface POUpdate {
  po_ref: string
  evidence: string
  confidence: number
  line_updates: LineUpdate[]
}

interface LLMOutput {
  po_updates: POUpdate[]
  unmatched_mentions: string[]
}

interface MatchSummary {
  proposed: number
  unmatched_pos: string[]
  unmatched_skus: Array<{ po_ref: string; sku_or_code: string }>
}

// ── normalization ────────────────────────────────────────────────────────────

function normalizeRef(s: string): string {
  return (
    s
      .toUpperCase()
      // Strip common "PO" prefix when followed by a separator or digit
      // so "PO-12" ~ "12" ~ "PO12" but "POMADE" is left alone
      .replace(/^PO(?=[-_\s]|\d)/, '')
      .replace(/^[-_\s]+/, '')   // strip any leading separator left behind
      .replace(/[\s\-_]/g, '')   // strip remaining separators
      .replace(/0+(\d)/g, '$1')  // collapse leading zeros in digit runs
  )
}

// ── PO resolution ────────────────────────────────────────────────────────────

interface PORow { id: string; version: number; reference_num: string }

async function resolvePO(
  db: pg.PoolClient,
  supplierId: string,
  poRef: string,
): Promise<{ row: PORow; match_confidence: number } | null> {
  // 1. Exact match
  const exact = await db.query<PORow>(
    `SELECT id, version, reference_num
     FROM purchase_order
     WHERE supplier_id = $1 AND reference_num = $2`,
    [supplierId, poRef],
  )
  if (exact.rows.length) return { row: exact.rows[0], match_confidence: 1.0 }

  // 2. Normalized match — load all POs for supplier and compare in memory
  const all = await db.query<PORow>(
    `SELECT id, version, reference_num FROM purchase_order WHERE supplier_id = $1`,
    [supplierId],
  )
  const norm = normalizeRef(poRef)
  const hit = all.rows.find((r) => normalizeRef(r.reference_num) === norm)
  if (hit) return { row: hit, match_confidence: 0.9 }

  return null
}

// ── line resolution ──────────────────────────────────────────────────────────

interface LineRow {
  id: string
  version: number
  quantity: string
  delivery_date: Date | null
  sku: string
}

async function resolveLine(
  db: pg.PoolClient,
  poId: string,
  supplierId: string,
  skuOrCode: string,
): Promise<{ row: LineRow; match_confidence: number } | null> {
  // 1. Exact SKU match on lines of this PO
  const exact = await db.query<LineRow>(
    `SELECT pol.id, pol.version, pol.quantity, pol.delivery_date, p.sku
     FROM   purchase_order_line pol
     JOIN   product p ON p.id = pol.product_id
     WHERE  pol.purchase_order_id = $1 AND p.sku = $2
     LIMIT  1`,
    [poId, skuOrCode],
  )
  if (exact.rows.length) return { row: exact.rows[0], match_confidence: 1.0 }

  // 2. Supplier SKU alias match via supplier_product
  const alias = await db.query<LineRow>(
    `SELECT pol.id, pol.version, pol.quantity, pol.delivery_date, p.sku
     FROM   purchase_order_line pol
     JOIN   product p            ON p.id = pol.product_id
     JOIN   supplier_product sp  ON sp.product_id = p.id
     WHERE  pol.purchase_order_id = $1
       AND  sp.supplier_id = $2
       AND  sp.supplier_sku = $3
     LIMIT  1`,
    [poId, supplierId, skuOrCode],
  )
  if (alias.rows.length) return { row: alias.rows[0], match_confidence: 0.9 }

  return null
}

// ── old value snapshot ───────────────────────────────────────────────────────

function getOldValue(line: LineRow, field: string): string | null {
  switch (field) {
    case 'delivery_date':
      return line.delivery_date ? line.delivery_date.toISOString().slice(0, 10) : null
    case 'quantity':
      return String(line.quantity)
    default:
      return null
  }
}

// ── core match function ──────────────────────────────────────────────────────

export async function match(runId: string): Promise<MatchSummary> {
  const db = await pool.connect()
  try {
    const runRes = await db.query<{ email_id: string; llm_output: LLMOutput }>(
      `SELECT email_id, llm_output FROM extraction_runs WHERE id = $1`,
      [runId],
    )
    if (!runRes.rows.length) throw new Error(`extraction_run not found: ${runId}`)
    const { email_id: emailId, llm_output: llmOutput } = runRes.rows[0]

    // Resolve supplier from email sender
    const emailRes = await db.query<{ sender: string }>(
      `SELECT sender FROM emails WHERE id = $1`,
      [emailId],
    )
    const sender = emailRes.rows[0]?.sender ?? ''
    const addrMatch = sender.match(/<([^>]+)>/)
    const senderEmail = (addrMatch ? addrMatch[1] : sender).toLowerCase()

    const supplierRes = await db.query<{ id: string }>(
      `SELECT s.id FROM supplier s WHERE lower(s.email) = $1
       UNION
       SELECT s.id FROM supplier s
       JOIN supplier_email_aliases sea ON sea.supplier_id = s.id
       WHERE lower(sea.email_address) = $1
       LIMIT 1`,
      [senderEmail],
    )
    const supplierId = supplierRes.rows[0]?.id ?? null

    const summary: MatchSummary = { proposed: 0, unmatched_pos: [], unmatched_skus: [] }

    await db.query('BEGIN')

    for (const poUpdate of llmOutput.po_updates ?? []) {
      if (!supplierId) {
        summary.unmatched_pos.push(poUpdate.po_ref)
        continue
      }

      const poMatch = await resolvePO(db, supplierId, poUpdate.po_ref)
      if (!poMatch) {
        summary.unmatched_pos.push(poUpdate.po_ref)
        continue
      }

      for (const lu of poUpdate.line_updates) {
        const lineMatch = await resolveLine(db, poMatch.row.id, supplierId, lu.sku_or_code)
        if (!lineMatch) {
          summary.unmatched_skus.push({ po_ref: poUpdate.po_ref, sku_or_code: lu.sku_or_code })
          continue
        }

        const extractionConfidence = lu.confidence
        const matchConfidence      = poMatch.match_confidence * lineMatch.match_confidence
        const combinedConfidence   = extractionConfidence * matchConfidence

        await db.query(
          `INSERT INTO proposed_changes
             (email_id, extraction_run_id,
              target_table, target_record_id, target_record_version,
              field_name, old_value, new_value,
              evidence_text,
              extraction_confidence, match_confidence, combined_confidence,
              status)
           VALUES ($1,$2,'purchase_order_line',$3,$4,$5,$6,$7,$8,$9,$10,$11,'pending')`,
          [
            emailId, runId,
            lineMatch.row.id, lineMatch.row.version,
            lu.field,
            getOldValue(lineMatch.row, lu.field),
            lu.new_value,
            lu.evidence,
            extractionConfidence, matchConfidence, combinedConfidence,
          ],
        )
        summary.proposed++
      }
    }

    const anyUnmatched =
      summary.unmatched_pos.length > 0 || summary.unmatched_skus.length > 0
    await db.query(
      `UPDATE emails SET status = $1 WHERE id = $2`,
      [anyUnmatched ? 'needs_review' : 'matched', emailId],
    )

    await db.query('COMMIT')
    return summary
  } catch (err) {
    await db.query('ROLLBACK').catch(() => {})
    throw err
  } finally {
    db.release()
  }
}

// ── CLI entrypoint ───────────────────────────────────────────────────────────

import { fileURLToPath } from 'url'
if (fileURLToPath(import.meta.url) === process.argv[1]) {
  const arg = process.argv.slice(2).find((a) => a !== '--')
  if (!arg) {
    console.error('usage: tsx backend/match.ts <extraction_run_uuid>')
    process.exit(1)
  }

  match(arg)
    .then((s) => {
      console.log(`proposed: ${s.proposed}`)
      if (s.unmatched_pos.length)
        console.log(`unmatched POs: ${s.unmatched_pos.join(', ')}`)
      if (s.unmatched_skus.length)
        console.log(
          `unmatched SKUs: ${s.unmatched_skus.map((u) => `${u.po_ref}/${u.sku_or_code}`).join(', ')}`,
        )
    })
    .catch((e) => {
      console.error(e)
      process.exit(1)
    })
}

File: backend/match.py
"""Matching step: extraction_run_id → proposed_changes rows.

Resolves the `po_ref` and `sku_or_code` strings the model produced back to
canonical database records, scores how confident the whole chain is, and
stages one `proposed_changes` row per line update.

Confidence is a product of two independent things:

    match_confidence    = po_match_confidence × line_match_confidence
    combined_confidence = extraction_confidence × match_confidence

where an exact reference/SKU hit scores 1.0 and a normalised or
supplier-alias hit scores 0.9. Anything at or above AUTO_APPLY_THRESHOLD is
staged as `approved` and written back immediately; everything else is left
`pending` for a human. Because the prompt only ever emits extraction
confidences of 1.0, 0.8 or 0.6, the default threshold of 0.95 means
auto-apply requires an unambiguous statement matched to an exact PO
reference and an exact SKU — any inference or alias hop routes to review.
"""

from __future__ import annotations

import logging
import re
from typing import Any, TypedDict

from psycopg import AsyncConnection

from backend.config import get_settings
from backend.db import connection
from backend.suppliers import resolve_supplier
from backend.writeback import apply_proposed_change

log = logging.getLogger(__name__)


class UnmatchedSku(TypedDict):
    po_ref: str
    sku_or_code: str


class MatchSummary(TypedDict):
    proposed: int
    auto_applied: int
    unmatched_pos: list[str]
    unmatched_skus: list[UnmatchedSku]


class ExtractionRunNotFoundError(LookupError):
    pass


# ── normalization ────────────────────────────────────────────────────────────

_PO_PREFIX = re.compile(r"^PO(?=[-_\s]|\d)")
_LEADING_SEP = re.compile(r"^[-_\s]+")
_SEPARATORS = re.compile(r"[\s\-_]")
_ZERO_RUN = re.compile(r"0+(\d)")


def normalize_ref(s: str) -> str:
    """Loosen a PO reference so "PO-12", "12" and "PO12" compare equal.

    Ported verbatim from the TypeScript matcher, including the fact that the
    final rule collapses zero runs anywhere in the string rather than only
    leading ones — so "PO-010" and "PO-10" both normalise to "1".
    """
    s = s.upper()
    # Strip common "PO" prefix when followed by a separator or digit
    # so "PO-12" ~ "12" ~ "PO12" but "POMADE" is left alone
    s = _PO_PREFIX.sub("", s)
    s = _LEADING_SEP.sub("", s)  # strip any leading separator left behind
    s = _SEPARATORS.sub("", s)  # strip remaining separators
    return _ZERO_RUN.sub(r"\1", s)  # collapse leading zeros in digit runs


# ── PO resolution ────────────────────────────────────────────────────────────


async def resolve_po(
    conn: AsyncConnection,
    supplier_id: str,
    po_ref: str,
) -> tuple[dict[str, Any], float] | None:
    """Resolve a quoted PO reference to a purchase_order row."""
    # 1. Exact match
    cur = await conn.execute(
        """
        SELECT id, version, reference_num
        FROM purchase_order
        WHERE supplier_id = %s AND reference_num = %s
        """,
        (supplier_id, po_ref),
    )
    exact = await cur.fetchone()
    if exact is not None:
        return exact, 1.0

    # 2. Normalized match — load all POs for supplier and compare in memory
    cur = await conn.execute(
        "SELECT id, version, reference_num FROM purchase_order WHERE supplier_id = %s",
        (supplier_id,),
    )
    norm = normalize_ref(po_ref)
    for row in await cur.fetchall():
        if normalize_ref(row["reference_num"]) == norm:
            return row, 0.9

    return None


# ── line resolution ──────────────────────────────────────────────────────────


async def resolve_line(
    conn: AsyncConnection,
    po_id: str,
    supplier_id: str,
    sku_or_code: str,
) -> tuple[dict[str, Any], float] | None:
    """Resolve a quoted product code to a line on this purchase order."""
    # 1. Exact SKU match on lines of this PO
    cur = await conn.execute(
        """
        SELECT pol.id, pol.version, pol.quantity, pol.delivery_date, p.sku
        FROM   purchase_order_line pol
        JOIN   product p ON p.id = pol.product_id
        WHERE  pol.purchase_order_id = %s AND p.sku = %s
        LIMIT  1
        """,
        (po_id, sku_or_code),
    )
    exact = await cur.fetchone()
    if exact is not None:
        return exact, 1.0

    # 2. Supplier SKU alias match via supplier_product
    cur = await conn.execute(
        """
        SELECT pol.id, pol.version, pol.quantity, pol.delivery_date, p.sku
        FROM   purchase_order_line pol
        JOIN   product p            ON p.id = pol.product_id
        JOIN   supplier_product sp  ON sp.product_id = p.id
        WHERE  pol.purchase_order_id = %s
          AND  sp.supplier_id = %s
          AND  sp.supplier_sku = %s
        LIMIT  1
        """,
        (po_id, supplier_id, sku_or_code),
    )
    alias = await cur.fetchone()
    if alias is not None:
        return alias, 0.9

    return None


# ── old value snapshot ───────────────────────────────────────────────────────


def get_old_value(line: dict[str, Any], field: str) -> str | None:
    """The canonical value at proposal time, shown to the reviewer as 'before'."""
    if field == "delivery_date":
        return line["delivery_date"].isoformat() if line["delivery_date"] else None
    if field == "quantity":
        return str(line["quantity"])
    return None


# ── core match function ──────────────────────────────────────────────────────


async def match(run_id: str, threshold: float | None = None) -> MatchSummary:
    """Stage proposed changes for one extraction run, auto-applying the safe ones."""
    if threshold is None:
        threshold = get_settings().auto_apply_threshold

    summary: MatchSummary = {
        "proposed": 0,
        "auto_applied": 0,
        "unmatched_pos": [],
        "unmatched_skus": [],
    }
    auto_apply_ids: list[str] = []

    async with connection() as conn:
        async with conn.transaction():
            cur = await conn.execute(
                "SELECT email_id, llm_output FROM extraction_runs WHERE id = %s",
                (run_id,),
            )
            run = await cur.fetchone()
            if run is None:
                raise ExtractionRunNotFoundError(f"extraction_run not found: {run_id}")

            email_id = run["email_id"]
            llm_output = run["llm_output"] or {}

            # Resolve supplier from email sender
            cur = await conn.execute(
                "SELECT sender FROM emails WHERE id = %s", (email_id,)
            )
            email_row = await cur.fetchone()
            sender = email_row["sender"] if email_row else ""

            supplier = await resolve_supplier(conn, sender)
            supplier_id = supplier["id"] if supplier else None

            for po_update in llm_output.get("po_updates") or []:
                if supplier_id is None:
                    summary["unmatched_pos"].append(po_update["po_ref"])
                    continue

                po_match = await resolve_po(conn, supplier_id, po_update["po_ref"])
                if po_match is None:
                    summary["unmatched_pos"].append(po_update["po_ref"])
                    continue
                po_row, po_confidence = po_match

                for lu in po_update.get("line_updates") or []:
                    line_match = await resolve_line(
                        conn, po_row["id"], supplier_id, lu["sku_or_code"]
                    )
                    if line_match is None:
                        summary["unmatched_skus"].append(
                            {
                                "po_ref": po_update["po_ref"],
                                "sku_or_code": lu["sku_or_code"],
                            }
                        )
                        continue
                    line_row, line_confidence = line_match

                    extraction_confidence = lu["confidence"]
                    match_confidence = po_confidence * line_confidence
                    combined_confidence = extraction_confidence * match_confidence
                    will_auto_apply = combined_confidence >= threshold

                    cur = await conn.execute(
                        """
                        INSERT INTO proposed_changes
                          (email_id, extraction_run_id,
                           target_table, target_record_id, target_record_version,
                           field_name, old_value, new_value,
                           evidence_text,
                           extraction_confidence, match_confidence, combined_confidence,
                           status)
                        VALUES (%s,%s,'purchase_order_line',%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
                        RETURNING id
                        """,
                        (
                            email_id,
                            run_id,
                            line_row["id"],
                            line_row["version"],
                            lu["field"],
                            get_old_value(line_row, lu["field"]),
                            lu["new_value"],
                            lu["evidence"],
                            extraction_confidence,
                            match_confidence,
                            combined_confidence,
                            "approved" if will_auto_apply else "pending",
                        ),
                    )
                    inserted = await cur.fetchone()
                    assert inserted is not None
                    summary["proposed"] += 1
                    if will_auto_apply:
                        auto_apply_ids.append(str(inserted["id"]))

            any_unmatched = bool(summary["unmatched_pos"] or summary["unmatched_skus"])
            await conn.execute(
                "UPDATE emails SET status = %s WHERE id = %s",
                ("needs_review" if any_unmatched else "matched", email_id),
            )

    # Run auto-apply after the match transaction commits; each apply is its
    # own transaction, so one failure cannot roll back the staged proposals.
    for pc_id in auto_apply_ids:
        try:
            result = await apply_proposed_change(pc_id, "auto")
            if result["status"] == "applied":
                summary["auto_applied"] += 1
        except Exception as err:
            log.error("auto-apply failed for %s: %s", pc_id, err)

    return summary
